addons/pos_stripe/models/pos_payment_method.py

```diff
@@ -0,0 +1,92 @@
+# coding: utf-8
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import logging
+import requests
+from werkzeug.urls import url_quote
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError, UserError
+
+_logger = logging.getLogger(__name__)
+TIMEOUT = 10
+
+class PosPaymentMethod(models.Model):
+    _inherit = 'pos.payment.method'
+
+    def _get_payment_terminal_selection(self):
+        return super()._get_payment_terminal_selection() + [('stripe', 'Stripe')]
+
+    # Stripe
+    stripe_serial_number = fields.Char(help='[Serial number of the stripe terminal], for example: WSC513105011295', copy=False)
+
+    @api.constrains('stripe_serial_number')
+    def _check_stripe_serial_number(self):
+        if not self.sudo()._get_stripe_secret_key():
+            raise ValidationError(_('Complete the Stripe onboarding.'))
+
+        for payment_method in self:
+            if not payment_method.stripe_serial_number:
+                continue
+            existing_payment_method = self.search([('id', '!=', payment_method.id),
+                                                   ('stripe_serial_number', '=', payment_method.stripe_serial_number)],
+                                                  limit=1)
+            if existing_payment_method:
+                raise ValidationError(_('Terminal %s is already used on payment method %s.',\
+                     payment_method.stripe_serial_number, existing_payment_method.display_name))
+
+
+    @api.model
+    def _get_stripe_secret_key(self):
+        stripe_secret_key = self.env['payment.acquirer'].search([('provider', '=', 'stripe')], limit=1).stripe_secret_key
+
+        if not stripe_secret_key:
+            raise ValidationError(_('Complete the Stripe onboarding.'))
+
+        return stripe_secret_key
+
+    @api.model
+    def stripe_connection_token(self):
+        endpoint = 'https://api.stripe.com/v1/terminal/connection_tokens'
+
+        try:
+            req = requests.post(endpoint, auth=(self.sudo()._get_stripe_secret_key(), ''), timeout=TIMEOUT)
+        except requests.exceptions.RequestException:
+            raise UserError(_("There are some issues between us and Stripe, try again later."))
+
+        if req.status_code is not requests.codes.ok:
+            raise UserError(_("There are some issues to contact Stripe."))
+
+        return req.json()
+
+    def stripe_payment_intent(self, amount):
+        # For Terminal payments, the 'payment_method_types' parameter must include
+        # 'card_present' and the 'capture_method' must be set to 'manual'
+        endpoint = 'https://api.stripe.com/v1/payment_intents'
+        currency = self.journal_id.currency_id or self.company_id.currency_id
+
+        try:
+            data = ("currency=%s&amount=%s&payment_method_types[]=card_present&capture_method=manual") % \
+                (currency.name, int(amount/currency.rounding))
```

Comment:
it's great that `currency` isn't injectable anymore :+1: 

But still you shouldn't be hand-crafting the parameters. It will be more readable and future-proof to use `werkzeug.urls.url_encode()` or `urllib.parse.urlencode(doseq=True)`

```suggestion
            data = werkzeug.urls.url_encode({
                "currency": currency.name,
                "amount": int(amount/currency.rounding),
                "payment_method_types[]": ["card_present"],
                "capture_method": "manual",
            })
```

Note that [`urllib.parse.urlencode`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) needs `doseq=True` to handle array values, whereas `werkzeug.urls.url_encode()` will do it by default. But you don't really care because you're only passing a single value, i.e. you could just use `"card_present"` instead of `["card_present"]`.